core/state/multi_tx_snapshot_test.go

```diff
@@ -0,0 +1,401 @@
+package state
```

Comment:
Agreed, I have added multiple tests with fuzzing, please let me know if any other tests should be added

Reply:
Thinking out loud: maybe we could set a "full debug" mode for multi tx commits that compare the multi tx apply and revert with apply and revert using state copies? Letting something like that run on the mainnet for a while could be a decent way to ensure the behaviour is correct

Reply:
> multi tx commits that compare the multi tx apply and revert with apply and revert using state copies

I had it running like this for about a week. There was no difference once we fixed several bugs. I also reproduced pretty much this scenario in fuzz tests. 

I didn't think it was worth merging in the code, but I am fairly confident we have correct behavior. 

If there is concern still I can add the code back to this branch, I just didn't because it was quite ugly and hacky.